{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Assignment 3 in System Genetics - Gil Sasson & Tomer Schweid",
   "id": "a7393e2afa31c893"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:27.593954Z",
     "start_time": "2024-09-20T14:41:27.326820Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Importing libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import GEOparse\n",
    "from sklearn.feature_selection import f_regression\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from itertools import chain\n",
    "from collections import Counter"
   ],
   "id": "118dec342b4ab535",
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Part A: eQTL analysis",
   "id": "c58417a6cd3deda6"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Preprocessing",
   "id": "9de105aad90805ba"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:27.596465Z",
     "start_time": "2024-09-20T14:41:27.594615Z"
    }
   },
   "cell_type": "code",
   "source": "directory_path = \"/Users/gilsasson/PycharmProjects/pythonProject/System_Genetics/data/\"",
   "id": "3fa2f8cad64a54ee",
   "outputs": [],
   "execution_count": 2
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:35.171143Z",
     "start_time": "2024-09-20T14:41:27.597166Z"
    }
   },
   "source": [
    "# Download the alphabet\n",
    "gse_myeloid = GEOparse.get_GEO(geo='GSE18067', destdir=directory_path)\n",
    "gse_liver = GEOparse.get_GEO(geo='GSE17522', destdir=directory_path)"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "20-Sep-2024 17:41:27 DEBUG utils - Directory /Users/gilsasson/PycharmProjects/pythonProject/System_Genetics/data/ already exists. Skipping.\n",
      "20-Sep-2024 17:41:27 INFO GEOparse - File already exist: using local version.\n",
      "20-Sep-2024 17:41:27 INFO GEOparse - Parsing /Users/gilsasson/PycharmProjects/pythonProject/System_Genetics/data/GSE18067_family.soft.gz: \n",
      "20-Sep-2024 17:41:27 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "20-Sep-2024 17:41:27 DEBUG GEOparse - SERIES: GSE18067\n",
      "20-Sep-2024 17:41:27 DEBUG GEOparse - PLATFORM: GPL6238\n",
      "20-Sep-2024 17:41:27 DEBUG GEOparse - SAMPLE: GSM451699\n",
      "20-Sep-2024 17:41:27 DEBUG GEOparse - SAMPLE: GSM451700\n",
      "20-Sep-2024 17:41:27 DEBUG GEOparse - SAMPLE: GSM451701\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451702\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451703\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451704\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451705\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451706\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451707\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451708\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451709\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451710\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451711\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451712\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451713\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451714\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451715\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451716\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451717\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451718\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451719\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451720\n",
      "20-Sep-2024 17:41:28 DEBUG GEOparse - SAMPLE: GSM451721\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451722\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451723\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451724\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451725\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451726\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451727\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451728\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451729\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451730\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451731\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451732\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451733\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451734\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451735\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451736\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451737\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451738\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451739\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451740\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451741\n",
      "20-Sep-2024 17:41:29 DEBUG GEOparse - SAMPLE: GSM451742\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451743\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451744\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451745\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451746\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451747\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451748\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451749\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451750\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451751\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451752\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451753\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451754\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451755\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451756\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451757\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451758\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451759\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451760\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451761\n",
      "20-Sep-2024 17:41:30 DEBUG GEOparse - SAMPLE: GSM451762\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451763\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451764\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451765\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451766\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451767\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451768\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451769\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451770\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451771\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451772\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451773\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451774\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451775\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451776\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451777\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451778\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451779\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451780\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451781\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451782\n",
      "20-Sep-2024 17:41:31 DEBUG GEOparse - SAMPLE: GSM451783\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451784\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451785\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451786\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451787\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451788\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451789\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451790\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451791\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451792\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451793\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451794\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451795\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451796\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451797\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM451798\n",
      "20-Sep-2024 17:41:32 DEBUG utils - Directory /Users/gilsasson/PycharmProjects/pythonProject/System_Genetics/data/ already exists. Skipping.\n",
      "20-Sep-2024 17:41:32 INFO GEOparse - File already exist: using local version.\n",
      "20-Sep-2024 17:41:32 INFO GEOparse - Parsing /Users/gilsasson/PycharmProjects/pythonProject/System_Genetics/data/GSE17522_family.soft.gz: \n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SERIES: GSE17522\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - PLATFORM: GPL6466\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM436705\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM436706\n",
      "20-Sep-2024 17:41:32 DEBUG GEOparse - SAMPLE: GSM436707\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436708\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436709\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436710\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436711\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436712\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436713\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436714\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436715\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436716\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436717\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436718\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436719\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436720\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436721\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436722\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436723\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436724\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436725\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436726\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436727\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436728\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436729\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436730\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436731\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436732\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436733\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436734\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436735\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436736\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436737\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436738\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436739\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436740\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436741\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436742\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436743\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436744\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436745\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436746\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436747\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436748\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436749\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436750\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436751\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436752\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436753\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436754\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436755\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436756\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436757\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436758\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436759\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436760\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436761\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436762\n",
      "20-Sep-2024 17:41:33 DEBUG GEOparse - SAMPLE: GSM436763\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436764\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436765\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436766\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436767\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436768\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436769\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436770\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436771\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436772\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436773\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436774\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436775\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436776\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436777\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436778\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436779\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436780\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436781\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436782\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436783\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436784\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436785\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436786\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436787\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436788\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436789\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436790\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436791\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436792\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436793\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436794\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436795\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436796\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436797\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436798\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436799\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436800\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436801\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436802\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436803\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436804\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436805\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436806\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436807\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436808\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436809\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436810\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436811\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436812\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436813\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436814\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436815\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436816\n",
      "20-Sep-2024 17:41:34 DEBUG GEOparse - SAMPLE: GSM436817\n",
      "20-Sep-2024 17:41:35 DEBUG GEOparse - SAMPLE: GSM436818\n",
      "20-Sep-2024 17:41:35 DEBUG GEOparse - SAMPLE: GSM436819\n",
      "20-Sep-2024 17:41:35 DEBUG GEOparse - SAMPLE: GSM436820\n",
      "20-Sep-2024 17:41:35 DEBUG GEOparse - SAMPLE: GSM436821\n",
      "20-Sep-2024 17:41:35 DEBUG GEOparse - SAMPLE: GSM436822\n",
      "20-Sep-2024 17:41:35 DEBUG GEOparse - SAMPLE: GSM436823\n",
      "20-Sep-2024 17:41:35 DEBUG GEOparse - SAMPLE: GSM436824\n",
      "20-Sep-2024 17:41:35 DEBUG GEOparse - SAMPLE: GSM436825\n",
      "20-Sep-2024 17:41:35 DEBUG GEOparse - SAMPLE: GSM436826\n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:35.185035Z",
     "start_time": "2024-09-20T14:41:35.172482Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Get the myeloid metadata\n",
    "metadata_myeloid = pd.read_csv(directory_path + \"metadata myeloid.txt\", sep=\"\\t\")\n",
    "metadata_myeloid.head()"
   ],
   "id": "9fb6a08ada2e33a2",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    GSM name                  Strain\n",
       "0  GSM451699    BXD13 Myeloid batch1\n",
       "1  GSM451700  BXD23 Erythroid batch1\n",
       "2  GSM451701  BXD40 Erythroid batch1\n",
       "3  GSM451702    BXD36 Myeloid batch1\n",
       "4  GSM451703        BXD6 Stem batch1"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM name</th>\n",
       "      <th>Strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSM451699</td>\n",
       "      <td>BXD13 Myeloid batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSM451700</td>\n",
       "      <td>BXD23 Erythroid batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSM451701</td>\n",
       "      <td>BXD40 Erythroid batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSM451702</td>\n",
       "      <td>BXD36 Myeloid batch1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSM451703</td>\n",
       "      <td>BXD6 Stem batch1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:35.190712Z",
     "start_time": "2024-09-20T14:41:35.185875Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Get the liver metadata\n",
    "metadata_liver = pd.read_csv(directory_path + \"metadata liver.txt\", sep=\"\\t\")\n",
    "metadata_liver.head()"
   ],
   "id": "4e719067e8efed4e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    GSM name                   Strain\n",
       "0  GSM436705  Liver_C57BL6J_M_B1_rep1\n",
       "1  GSM436706  Liver_C57BL6J_M_B1_rep2\n",
       "2  GSM436707       Liver_C57BL6J_F_B1\n",
       "3  GSM436708         Liver_DBA2J_F_B1\n",
       "4  GSM436709        Liver_B6D2F1_M_B1"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM name</th>\n",
       "      <th>Strain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSM436705</td>\n",
       "      <td>Liver_C57BL6J_M_B1_rep1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSM436706</td>\n",
       "      <td>Liver_C57BL6J_M_B1_rep2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSM436707</td>\n",
       "      <td>Liver_C57BL6J_F_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSM436708</td>\n",
       "      <td>Liver_DBA2J_F_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSM436709</td>\n",
       "      <td>Liver_B6D2F1_M_B1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:35.205939Z",
     "start_time": "2024-09-20T14:41:35.191571Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Get the gene identifiers for the myeloid alphabet\n",
    "gene_identifiers_df_myeloid = pd.DataFrame(gse_myeloid.gpls[list(gse_myeloid.gpls.keys())[0]].table)[[\"ID\" , \"Symbol\"]]\n",
    "gene_identifiers_df_myeloid.head()"
   ],
   "id": "6c339fa73cb94b7f",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                   ID         Symbol\n",
       "0       GI_38090455-S      LOC382362\n",
       "1  scl0011717.2_283-S          Ampd3\n",
       "2  scl52892.1.1_103-S  D830016O14Rik\n",
       "3    scl0012946.2_3-S           Crry\n",
       "4    scl18939.11_88-S            Ehf"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GI_38090455-S</td>\n",
       "      <td>LOC382362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scl0011717.2_283-S</td>\n",
       "      <td>Ampd3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scl52892.1.1_103-S</td>\n",
       "      <td>D830016O14Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scl0012946.2_3-S</td>\n",
       "      <td>Crry</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scl18939.11_88-S</td>\n",
       "      <td>Ehf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:35.212438Z",
     "start_time": "2024-09-20T14:41:35.206654Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Get the gene identifiers for the liver alphabet\n",
    "gene_identifiers_df_liver= pd.DataFrame(gse_liver.gpls[list(gse_liver.gpls.keys())[0]].table)[[\"ID\" , \"GENE_NAME\"]]\n",
    "gene_identifiers_df_liver.head()"
   ],
   "id": "807217ec8453d364",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "             ID                                          GENE_NAME\n",
       "0  A_51_P100021  human immunodeficiency virus type I enhancer b...\n",
       "1  A_51_P100034                            MIF4G domain containing\n",
       "2  A_51_P100052                SLIT and NTRK-like family, member 2\n",
       "3  A_51_P100065                         ligand of numb-protein X 1\n",
       "4  A_51_P100084                                                NaN"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GENE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_51_P100021</td>\n",
       "      <td>human immunodeficiency virus type I enhancer b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_51_P100034</td>\n",
       "      <td>MIF4G domain containing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_51_P100052</td>\n",
       "      <td>SLIT and NTRK-like family, member 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_51_P100065</td>\n",
       "      <td>ligand of numb-protein X 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_51_P100084</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:35.237734Z",
     "start_time": "2024-09-20T14:41:35.213149Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Get the genotype alphabet\n",
    "genotype_df = pd.read_csv(directory_path + \"genotypes.txt\")\n",
    "genotype_df.head()"
   ],
   "id": "48c4e0390343ebf2",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   0      Locus  Chr_Build37  Build37_position  BXD1  BXD2  BXD5  BXD6  BXD8  \\\n",
       "0  1  rs6269442            1           3482276     0     0     2     2     2   \n",
       "1  2  rs6365999            1           4811063     0     0     2     2     2   \n",
       "2  3  rs6376963            1           5008090     0     0     2     2     2   \n",
       "3  4  rs3677817            1           5176059     0     0     2     2     2   \n",
       "4  5  rs8236463            1           5579194     0     0     2     2     2   \n",
       "\n",
       "   BXD9  ...  BXD94  BXD95  BXD96  BXD97  BXD98  BXD99  BXD100  BXD101  \\\n",
       "0     0  ...      2      2      0      0      0      0       0       U   \n",
       "1     0  ...      2      2      0      0      0      0       0       U   \n",
       "2     0  ...      2      2      0      0      0      0       0       U   \n",
       "3     0  ...      2      2      0      0      0      0       0       U   \n",
       "4     0  ...      2      2      0      0      0      0       0       U   \n",
       "\n",
       "   BXD102  BXD103  \n",
       "0       U       U  \n",
       "1       U       U  \n",
       "2       U       U  \n",
       "3       U       U  \n",
       "4       U       U  \n",
       "\n",
       "[5 rows x 96 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>Locus</th>\n",
       "      <th>Chr_Build37</th>\n",
       "      <th>Build37_position</th>\n",
       "      <th>BXD1</th>\n",
       "      <th>BXD2</th>\n",
       "      <th>BXD5</th>\n",
       "      <th>BXD6</th>\n",
       "      <th>BXD8</th>\n",
       "      <th>BXD9</th>\n",
       "      <th>...</th>\n",
       "      <th>BXD94</th>\n",
       "      <th>BXD95</th>\n",
       "      <th>BXD96</th>\n",
       "      <th>BXD97</th>\n",
       "      <th>BXD98</th>\n",
       "      <th>BXD99</th>\n",
       "      <th>BXD100</th>\n",
       "      <th>BXD101</th>\n",
       "      <th>BXD102</th>\n",
       "      <th>BXD103</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs6269442</td>\n",
       "      <td>1</td>\n",
       "      <td>3482276</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>rs6365999</td>\n",
       "      <td>1</td>\n",
       "      <td>4811063</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>rs6376963</td>\n",
       "      <td>1</td>\n",
       "      <td>5008090</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>rs3677817</td>\n",
       "      <td>1</td>\n",
       "      <td>5176059</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>rs8236463</td>\n",
       "      <td>1</td>\n",
       "      <td>5579194</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 96 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:35.250080Z",
     "start_time": "2024-09-20T14:41:35.242823Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Define useful functions\n",
    "def parse_gse(gse):\n",
    "  final_df = pd.DataFrame(columns=[\"GSM_name\", \"ID_REF\", \"VALUE\"])\n",
    "  sample_dfs = []\n",
    "\n",
    "  # Iterate through each sample\n",
    "  for gsm_name, gsm in gse.gsms.items():\n",
    "      sample_data = {\n",
    "          \"GSM_name\": gsm_name,\n",
    "          \"ID_REF\": gsm.table[\"ID_REF\"],\n",
    "          \"VALUE\": gsm.table[\"VALUE\"],\n",
    "      }\n",
    "      sample_df = pd.DataFrame(sample_data)\n",
    "      sample_dfs.append(sample_df)\n",
    "  final_df = pd.concat(sample_dfs, ignore_index=True)\n",
    "  return final_df\n",
    "\n",
    "def merge(df1, df2, left_on, right_on):\n",
    "    df1 = df1.merge(df2, left_on=left_on, right_on=right_on)\n",
    "    return df1.drop(columns=[right_on, left_on])\n",
    "\n",
    "def filter_samples(df, cell_type=\"Myeloid\"):\n",
    "    # Extract relevant cell types\n",
    "    df = df[df[\"Strain\"].str.contains(cell_type)]\n",
    "    # Drop the technical replicates\n",
    "    df = df[~df[\"Strain\"].str.contains(\"Technical\")]\n",
    "    df = df[~df[\"Strain\"].str.contains(\"rep\")]\n",
    "    return df\n",
    "\n",
    "def pivot_strains(df_merged, index):\n",
    "    # Pivot the alphabet\n",
    "    df_pivot = df_merged.pivot_table(columns=\"Strain\", values=\"VALUE\", index=index, aggfunc=\"mean\")\n",
    "    return df_pivot\n",
    "\n",
    "def mean_batches(df, cell_type=\"Myeloid\"):\n",
    "    if cell_type == \"Myeloid\":\n",
    "        batch_names = [\"a\", \"b\"]\n",
    "        # Extract strain names from column names\n",
    "        df.columns = pd.MultiIndex.from_tuples([col.split(' ', 1) for col in df.columns], names=['strain', 'batch'])\n",
    "        # Calculate the mean across batches for each strain\n",
    "        mean_df = df.T.groupby(level='strain').mean()\n",
    "        return mean_df.T\n",
    "    elif cell_type == \"Liver\":\n",
    "        # Mean across sex for each strain\n",
    "        df.columns = pd.MultiIndex.from_tuples([col.split('_', 2) for col in df.columns],names=['liver','strain','sex'])\n",
    "        # Calculate the mean across sex for each strain\n",
    "        mean_df = df.T.groupby(level=['strain']).mean()\n",
    "        return mean_df.T\n",
    "    \n",
    "\n",
    "def mean_strains(df_pivot):\n",
    "        for strain in df_pivot.columns:\n",
    "            if strain[-1].lower() == 'a':\n",
    "                strain_name = strain[:-1]\n",
    "                if strain_name + 'b' in df_pivot.columns:\n",
    "                    df_pivot[strain_name] = df_pivot[[strain, strain_name + 'b']].mean(axis=1)\n",
    "                    df_pivot = df_pivot.drop(columns=[strain, strain_name + 'b'])\n",
    "                else:\n",
    "                    df_pivot[strain_name] = df_pivot[strain]\n",
    "                    df_pivot = df_pivot.drop(columns=[strain])\n",
    "            elif strain[-1].lower() == 'b' :\n",
    "                strain_name = strain[:-1]\n",
    "                if strain_name + 'a' in df_pivot.columns:\n",
    "                    df_pivot[strain_name] = df_pivot[[strain, strain_name + 'a']].mean(axis=1)\n",
    "                    df_pivot = df_pivot.drop(columns=[strain, strain_name + 'a'])\n",
    "        return df_pivot"
   ],
   "id": "6d0f34c6f802b222",
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:39.079666Z",
     "start_time": "2024-09-20T14:41:35.252325Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Get the myeloid alphabet\n",
    "myeloid_df = parse_gse(gse_myeloid)\n",
    "# Merge the metadata with the alphabet\n",
    "myeloid_df = merge(myeloid_df, metadata_myeloid, \"GSM_name\", \"GSM name\")\n",
    "# Merge the gene identifiers with the alphabet\n",
    "myeloid_df = merge(myeloid_df, gene_identifiers_df_myeloid, \"ID_REF\", \"ID\")\n",
    "# Filter the samples to include only the myeloid cells alphabet\n",
    "myeloid_df = filter_samples(myeloid_df)\n",
    "# Reshape the alphabet\n",
    "myeloid_df = pivot_strains(myeloid_df, \"Symbol\")\n",
    "# Calculate the mean across batches for each strain\n",
    "myeloid_df = mean_batches(myeloid_df)\n",
    "# Merge duplicate strains columns\n",
    "myeloid_df = mean_strains(myeloid_df)\n",
    "myeloid_df.head()"
   ],
   "id": "7537a9c2ed96625a",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "strain             BXD1     BXD11     BXD12     BXD13     BXD14     BXD16  \\\n",
       "Symbol                                                                      \n",
       "0610005A07Rik  6.439252  6.418787  6.461313  6.459590  6.392369  6.515128   \n",
       "0610005C13Rik  6.462294  6.556949  6.591845  6.573851  6.482562  6.616155   \n",
       "0610005H09Rik  6.960774  7.096886  7.711693  7.647623  7.651882  7.734397   \n",
       "0610005I04     6.532020  6.594255  6.489756  6.697061  6.497515  6.548089   \n",
       "0610005K03Rik  7.661794  7.659268  7.786213  7.787414  8.152209  8.030581   \n",
       "\n",
       "strain            BXD18     BXD19      BXD2    BXD21B  ...    BXD32B  \\\n",
       "Symbol                                                 ...             \n",
       "0610005A07Rik  6.446004  6.392673  6.263298  6.318342  ...  6.391703   \n",
       "0610005C13Rik  6.494463  6.553485  6.492469  6.409468  ...  6.616821   \n",
       "0610005H09Rik  7.822844  7.631427  6.896540  7.810171  ...  8.154312   \n",
       "0610005I04     6.637054  6.616757  6.633698  6.774890  ...  6.711600   \n",
       "0610005K03Rik  7.836622  8.056267  7.666384  7.882478  ...  7.776629   \n",
       "\n",
       "strain           BXD34B     BXD36     BXD40     BXD42      BXD6      BXD8  \\\n",
       "Symbol                                                                      \n",
       "0610005A07Rik  6.414189  6.339669  6.244116  6.322840  6.480558  6.522814   \n",
       "0610005C13Rik  6.510064  6.530035  6.591437  6.526547  6.454843  6.496264   \n",
       "0610005H09Rik  7.292728  8.094586  8.252535  7.375777  7.562560  7.577809   \n",
       "0610005I04     6.488575  6.650941  6.705175  6.630276  6.831132  6.739256   \n",
       "0610005K03Rik  8.313372  7.908074  7.922678  7.836622  7.928899  7.988844   \n",
       "\n",
       "strain             BXD9     BXD28     BXD33  \n",
       "Symbol                                       \n",
       "0610005A07Rik  6.373272  6.529782  6.513522  \n",
       "0610005C13Rik  6.586568  6.604400  6.608433  \n",
       "0610005H09Rik  7.829545  7.516431  7.651062  \n",
       "0610005I04     6.606980  6.953100  6.804694  \n",
       "0610005K03Rik  7.882973  7.781489  7.780241  \n",
       "\n",
       "[5 rows x 24 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strain</th>\n",
       "      <th>BXD1</th>\n",
       "      <th>BXD11</th>\n",
       "      <th>BXD12</th>\n",
       "      <th>BXD13</th>\n",
       "      <th>BXD14</th>\n",
       "      <th>BXD16</th>\n",
       "      <th>BXD18</th>\n",
       "      <th>BXD19</th>\n",
       "      <th>BXD2</th>\n",
       "      <th>BXD21B</th>\n",
       "      <th>...</th>\n",
       "      <th>BXD32B</th>\n",
       "      <th>BXD34B</th>\n",
       "      <th>BXD36</th>\n",
       "      <th>BXD40</th>\n",
       "      <th>BXD42</th>\n",
       "      <th>BXD6</th>\n",
       "      <th>BXD8</th>\n",
       "      <th>BXD9</th>\n",
       "      <th>BXD28</th>\n",
       "      <th>BXD33</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005A07Rik</th>\n",
       "      <td>6.439252</td>\n",
       "      <td>6.418787</td>\n",
       "      <td>6.461313</td>\n",
       "      <td>6.459590</td>\n",
       "      <td>6.392369</td>\n",
       "      <td>6.515128</td>\n",
       "      <td>6.446004</td>\n",
       "      <td>6.392673</td>\n",
       "      <td>6.263298</td>\n",
       "      <td>6.318342</td>\n",
       "      <td>...</td>\n",
       "      <td>6.391703</td>\n",
       "      <td>6.414189</td>\n",
       "      <td>6.339669</td>\n",
       "      <td>6.244116</td>\n",
       "      <td>6.322840</td>\n",
       "      <td>6.480558</td>\n",
       "      <td>6.522814</td>\n",
       "      <td>6.373272</td>\n",
       "      <td>6.529782</td>\n",
       "      <td>6.513522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>6.462294</td>\n",
       "      <td>6.556949</td>\n",
       "      <td>6.591845</td>\n",
       "      <td>6.573851</td>\n",
       "      <td>6.482562</td>\n",
       "      <td>6.616155</td>\n",
       "      <td>6.494463</td>\n",
       "      <td>6.553485</td>\n",
       "      <td>6.492469</td>\n",
       "      <td>6.409468</td>\n",
       "      <td>...</td>\n",
       "      <td>6.616821</td>\n",
       "      <td>6.510064</td>\n",
       "      <td>6.530035</td>\n",
       "      <td>6.591437</td>\n",
       "      <td>6.526547</td>\n",
       "      <td>6.454843</td>\n",
       "      <td>6.496264</td>\n",
       "      <td>6.586568</td>\n",
       "      <td>6.604400</td>\n",
       "      <td>6.608433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610005H09Rik</th>\n",
       "      <td>6.960774</td>\n",
       "      <td>7.096886</td>\n",
       "      <td>7.711693</td>\n",
       "      <td>7.647623</td>\n",
       "      <td>7.651882</td>\n",
       "      <td>7.734397</td>\n",
       "      <td>7.822844</td>\n",
       "      <td>7.631427</td>\n",
       "      <td>6.896540</td>\n",
       "      <td>7.810171</td>\n",
       "      <td>...</td>\n",
       "      <td>8.154312</td>\n",
       "      <td>7.292728</td>\n",
       "      <td>8.094586</td>\n",
       "      <td>8.252535</td>\n",
       "      <td>7.375777</td>\n",
       "      <td>7.562560</td>\n",
       "      <td>7.577809</td>\n",
       "      <td>7.829545</td>\n",
       "      <td>7.516431</td>\n",
       "      <td>7.651062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610005I04</th>\n",
       "      <td>6.532020</td>\n",
       "      <td>6.594255</td>\n",
       "      <td>6.489756</td>\n",
       "      <td>6.697061</td>\n",
       "      <td>6.497515</td>\n",
       "      <td>6.548089</td>\n",
       "      <td>6.637054</td>\n",
       "      <td>6.616757</td>\n",
       "      <td>6.633698</td>\n",
       "      <td>6.774890</td>\n",
       "      <td>...</td>\n",
       "      <td>6.711600</td>\n",
       "      <td>6.488575</td>\n",
       "      <td>6.650941</td>\n",
       "      <td>6.705175</td>\n",
       "      <td>6.630276</td>\n",
       "      <td>6.831132</td>\n",
       "      <td>6.739256</td>\n",
       "      <td>6.606980</td>\n",
       "      <td>6.953100</td>\n",
       "      <td>6.804694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610005K03Rik</th>\n",
       "      <td>7.661794</td>\n",
       "      <td>7.659268</td>\n",
       "      <td>7.786213</td>\n",
       "      <td>7.787414</td>\n",
       "      <td>8.152209</td>\n",
       "      <td>8.030581</td>\n",
       "      <td>7.836622</td>\n",
       "      <td>8.056267</td>\n",
       "      <td>7.666384</td>\n",
       "      <td>7.882478</td>\n",
       "      <td>...</td>\n",
       "      <td>7.776629</td>\n",
       "      <td>8.313372</td>\n",
       "      <td>7.908074</td>\n",
       "      <td>7.922678</td>\n",
       "      <td>7.836622</td>\n",
       "      <td>7.928899</td>\n",
       "      <td>7.988844</td>\n",
       "      <td>7.882973</td>\n",
       "      <td>7.781489</td>\n",
       "      <td>7.780241</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:39.114124Z",
     "start_time": "2024-09-20T14:41:39.080771Z"
    }
   },
   "cell_type": "code",
   "source": [
    "relevant_strains = genotype_df.columns[4:]\n",
    "# Filter the relevant strains to include only those present in myeloid_df\n",
    "filtered_strains_myeloid = [col for col in relevant_strains if col in myeloid_df.columns]\n",
    "# Select the filtered columns from myeloid_df\n",
    "myeloid_df = myeloid_df.loc[:, filtered_strains_myeloid]\n",
    "# Get only the strains present in myeolid_df from the genotype alphabet\n",
    "representative_snps_myeloid = genotype_df[filtered_strains_myeloid]\n",
    "# Now look for snps that have exactly same values for all the strains\n",
    "duplicates = representative_snps_myeloid[representative_snps_myeloid.duplicated()]\n",
    "print(f\"The number of duplicate SNPs is: {len(duplicates)}\")\n",
    "# Remove the duplicate SNPs\n",
    "representative_snps_myeloid = representative_snps_myeloid.drop_duplicates()"
   ],
   "id": "8167f47da145fd1f",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of duplicate SNPs is: 3076\n"
     ]
    }
   ],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:39.148336Z",
     "start_time": "2024-09-20T14:41:39.114992Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Now we have the alphabet in the correct format\n",
    "# We can start filtering the alphabet\n",
    "# First, we will filter out the genes with no gene identifier\n",
    "myeloid_df_filtered = myeloid_df.dropna()\n",
    "# Then, we will filter out the genes with low variance\n",
    "row_variance = myeloid_df_filtered.var(axis=1)\n",
    "average_variance = row_variance.mean()\n",
    "print(f\"The average variance is: {round(average_variance, 3)}\")\n",
    "myeloid_df_filtered = myeloid_df_filtered[row_variance > average_variance/10]\n",
    "# Now we will filter out the genes with low maximum expression\n",
    "row_max = myeloid_df_filtered.max(axis=1)\n",
    "average_max = row_max.mean()\n",
    "print(f\"The average maximum expression value is: {round(average_max, 3)}\")\n",
    "myeloid_df_filtered = myeloid_df_filtered[row_max > average_max-1.2]\n",
    "print(f\"The number of genes before filtering is: {len(myeloid_df)} \\nThe number of genes after filtering is: {len(myeloid_df_filtered)}\")\n",
    "myeloid_df_filtered"
   ],
   "id": "88bac7e54099c5be",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The average variance is: 0.044\n",
      "The average maximum expression value is: 7.661\n",
      "The number of genes before filtering is: 33718 \n",
      "The number of genes after filtering is: 30122\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "strain                                         BXD1      BXD2      BXD6  \\\n",
       "Symbol                                                                    \n",
       "0610005A07Rik                              6.439252  6.263298  6.480558   \n",
       "0610005H09Rik                              6.960774  6.896540  7.562560   \n",
       "0610005I04                                 6.532020  6.633698  6.831132   \n",
       "0610005K03Rik                              7.661794  7.666384  7.928899   \n",
       "0610006F02Rik                              6.839912  6.820642  6.938233   \n",
       "...                                             ...       ...       ...   \n",
       "neuronal nitric-oxide synthase isoform mu  6.429438  6.674736  6.401588   \n",
       "p                                          6.541973  6.359801  6.517618   \n",
       "scavenger receptor                         7.106817  7.059044  6.899810   \n",
       "sty                                        8.787945  8.523423  7.741574   \n",
       "tea                                        6.494957  6.497032  6.438302   \n",
       "\n",
       "strain                                         BXD8      BXD9     BXD11  \\\n",
       "Symbol                                                                    \n",
       "0610005A07Rik                              6.522814  6.373272  6.418787   \n",
       "0610005H09Rik                              7.577809  7.829545  7.096886   \n",
       "0610005I04                                 6.739256  6.606980  6.594255   \n",
       "0610005K03Rik                              7.988844  7.882973  7.659268   \n",
       "0610006F02Rik                              6.882953  6.858111  6.752358   \n",
       "...                                             ...       ...       ...   \n",
       "neuronal nitric-oxide synthase isoform mu  6.363483  6.342921  6.566346   \n",
       "p                                          6.426932  6.535042  6.479159   \n",
       "scavenger receptor                         6.699562  7.049678  7.526465   \n",
       "sty                                        9.063488  8.008797  9.646043   \n",
       "tea                                        6.627027  6.473407  6.454031   \n",
       "\n",
       "strain                                        BXD12     BXD13     BXD14  \\\n",
       "Symbol                                                                    \n",
       "0610005A07Rik                              6.461313  6.459590  6.392369   \n",
       "0610005H09Rik                              7.711693  7.647623  7.651882   \n",
       "0610005I04                                 6.489756  6.697061  6.497515   \n",
       "0610005K03Rik                              7.786213  7.787414  8.152209   \n",
       "0610006F02Rik                              6.831553  6.860729  6.741501   \n",
       "...                                             ...       ...       ...   \n",
       "neuronal nitric-oxide synthase isoform mu  6.379514  6.349456  6.532331   \n",
       "p                                          6.212613  6.386862  6.508521   \n",
       "scavenger receptor                         7.402048  7.746785  6.798596   \n",
       "sty                                        9.472719  8.881412  9.161894   \n",
       "tea                                        6.418251  6.405716  6.471698   \n",
       "\n",
       "strain                                        BXD16  ...     BXD19     BXD23  \\\n",
       "Symbol                                               ...                       \n",
       "0610005A07Rik                              6.515128  ...  6.392673  6.343721   \n",
       "0610005H09Rik                              7.734397  ...  7.631427  7.658030   \n",
       "0610005I04                                 6.548089  ...  6.616757  6.640003   \n",
       "0610005K03Rik                              8.030581  ...  8.056267  7.920657   \n",
       "0610006F02Rik                              6.867071  ...  6.766329  6.935824   \n",
       "...                                             ...  ...       ...       ...   \n",
       "neuronal nitric-oxide synthase isoform mu  6.577870  ...  6.470454  6.517419   \n",
       "p                                          6.347702  ...  6.104731  6.359801   \n",
       "scavenger receptor                         8.034765  ...  6.715678  7.140271   \n",
       "sty                                        9.242461  ...  9.029700  9.350928   \n",
       "tea                                        6.501229  ...  6.535940  6.373899   \n",
       "\n",
       "strain                                        BXD27     BXD28     BXD29  \\\n",
       "Symbol                                                                    \n",
       "0610005A07Rik                              6.486555  6.529782  6.409871   \n",
       "0610005H09Rik                              7.765902  7.516431  7.733955   \n",
       "0610005I04                                 6.907307  6.953100  6.727989   \n",
       "0610005K03Rik                              7.900948  7.781489  7.869633   \n",
       "0610006F02Rik                              6.930366  6.943975  6.996387   \n",
       "...                                             ...       ...       ...   \n",
       "neuronal nitric-oxide synthase isoform mu  6.548269  6.485421  6.447770   \n",
       "p                                          6.466790  6.642090  6.369591   \n",
       "scavenger receptor                         6.874077  7.132313  7.834666   \n",
       "sty                                        9.545547  9.239431  9.492746   \n",
       "tea                                        6.444591  6.471909  6.478121   \n",
       "\n",
       "strain                                         BXD31     BXD33     BXD36  \\\n",
       "Symbol                                                                     \n",
       "0610005A07Rik                               6.494575  6.513522  6.339669   \n",
       "0610005H09Rik                               7.371241  7.651062  8.094586   \n",
       "0610005I04                                  6.721578  6.804694  6.650941   \n",
       "0610005K03Rik                               7.581931  7.780241  7.908074   \n",
       "0610006F02Rik                               6.960253  6.991211  6.953835   \n",
       "...                                              ...       ...       ...   \n",
       "neuronal nitric-oxide synthase isoform mu   6.461752  6.479435  6.402252   \n",
       "p                                           6.415152  6.336126  6.395560   \n",
       "scavenger receptor                          7.702109  6.960724  6.652632   \n",
       "sty                                        10.106197  9.241477  8.622802   \n",
       "tea                                         6.372046  6.570508  6.388176   \n",
       "\n",
       "strain                                        BXD40     BXD42  \n",
       "Symbol                                                         \n",
       "0610005A07Rik                              6.244116  6.322840  \n",
       "0610005H09Rik                              8.252535  7.375777  \n",
       "0610005I04                                 6.705175  6.630276  \n",
       "0610005K03Rik                              7.922678  7.836622  \n",
       "0610006F02Rik                              6.831518  7.015863  \n",
       "...                                             ...       ...  \n",
       "neuronal nitric-oxide synthase isoform mu  6.610253  6.398350  \n",
       "p                                          6.400445  6.347738  \n",
       "scavenger receptor                         7.252858  6.451716  \n",
       "sty                                        9.116669  9.550628  \n",
       "tea                                        6.506383  6.417805  \n",
       "\n",
       "[30122 rows x 21 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strain</th>\n",
       "      <th>BXD1</th>\n",
       "      <th>BXD2</th>\n",
       "      <th>BXD6</th>\n",
       "      <th>BXD8</th>\n",
       "      <th>BXD9</th>\n",
       "      <th>BXD11</th>\n",
       "      <th>BXD12</th>\n",
       "      <th>BXD13</th>\n",
       "      <th>BXD14</th>\n",
       "      <th>BXD16</th>\n",
       "      <th>...</th>\n",
       "      <th>BXD19</th>\n",
       "      <th>BXD23</th>\n",
       "      <th>BXD27</th>\n",
       "      <th>BXD28</th>\n",
       "      <th>BXD29</th>\n",
       "      <th>BXD31</th>\n",
       "      <th>BXD33</th>\n",
       "      <th>BXD36</th>\n",
       "      <th>BXD40</th>\n",
       "      <th>BXD42</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005A07Rik</th>\n",
       "      <td>6.439252</td>\n",
       "      <td>6.263298</td>\n",
       "      <td>6.480558</td>\n",
       "      <td>6.522814</td>\n",
       "      <td>6.373272</td>\n",
       "      <td>6.418787</td>\n",
       "      <td>6.461313</td>\n",
       "      <td>6.459590</td>\n",
       "      <td>6.392369</td>\n",
       "      <td>6.515128</td>\n",
       "      <td>...</td>\n",
       "      <td>6.392673</td>\n",
       "      <td>6.343721</td>\n",
       "      <td>6.486555</td>\n",
       "      <td>6.529782</td>\n",
       "      <td>6.409871</td>\n",
       "      <td>6.494575</td>\n",
       "      <td>6.513522</td>\n",
       "      <td>6.339669</td>\n",
       "      <td>6.244116</td>\n",
       "      <td>6.322840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610005H09Rik</th>\n",
       "      <td>6.960774</td>\n",
       "      <td>6.896540</td>\n",
       "      <td>7.562560</td>\n",
       "      <td>7.577809</td>\n",
       "      <td>7.829545</td>\n",
       "      <td>7.096886</td>\n",
       "      <td>7.711693</td>\n",
       "      <td>7.647623</td>\n",
       "      <td>7.651882</td>\n",
       "      <td>7.734397</td>\n",
       "      <td>...</td>\n",
       "      <td>7.631427</td>\n",
       "      <td>7.658030</td>\n",
       "      <td>7.765902</td>\n",
       "      <td>7.516431</td>\n",
       "      <td>7.733955</td>\n",
       "      <td>7.371241</td>\n",
       "      <td>7.651062</td>\n",
       "      <td>8.094586</td>\n",
       "      <td>8.252535</td>\n",
       "      <td>7.375777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610005I04</th>\n",
       "      <td>6.532020</td>\n",
       "      <td>6.633698</td>\n",
       "      <td>6.831132</td>\n",
       "      <td>6.739256</td>\n",
       "      <td>6.606980</td>\n",
       "      <td>6.594255</td>\n",
       "      <td>6.489756</td>\n",
       "      <td>6.697061</td>\n",
       "      <td>6.497515</td>\n",
       "      <td>6.548089</td>\n",
       "      <td>...</td>\n",
       "      <td>6.616757</td>\n",
       "      <td>6.640003</td>\n",
       "      <td>6.907307</td>\n",
       "      <td>6.953100</td>\n",
       "      <td>6.727989</td>\n",
       "      <td>6.721578</td>\n",
       "      <td>6.804694</td>\n",
       "      <td>6.650941</td>\n",
       "      <td>6.705175</td>\n",
       "      <td>6.630276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610005K03Rik</th>\n",
       "      <td>7.661794</td>\n",
       "      <td>7.666384</td>\n",
       "      <td>7.928899</td>\n",
       "      <td>7.988844</td>\n",
       "      <td>7.882973</td>\n",
       "      <td>7.659268</td>\n",
       "      <td>7.786213</td>\n",
       "      <td>7.787414</td>\n",
       "      <td>8.152209</td>\n",
       "      <td>8.030581</td>\n",
       "      <td>...</td>\n",
       "      <td>8.056267</td>\n",
       "      <td>7.920657</td>\n",
       "      <td>7.900948</td>\n",
       "      <td>7.781489</td>\n",
       "      <td>7.869633</td>\n",
       "      <td>7.581931</td>\n",
       "      <td>7.780241</td>\n",
       "      <td>7.908074</td>\n",
       "      <td>7.922678</td>\n",
       "      <td>7.836622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006F02Rik</th>\n",
       "      <td>6.839912</td>\n",
       "      <td>6.820642</td>\n",
       "      <td>6.938233</td>\n",
       "      <td>6.882953</td>\n",
       "      <td>6.858111</td>\n",
       "      <td>6.752358</td>\n",
       "      <td>6.831553</td>\n",
       "      <td>6.860729</td>\n",
       "      <td>6.741501</td>\n",
       "      <td>6.867071</td>\n",
       "      <td>...</td>\n",
       "      <td>6.766329</td>\n",
       "      <td>6.935824</td>\n",
       "      <td>6.930366</td>\n",
       "      <td>6.943975</td>\n",
       "      <td>6.996387</td>\n",
       "      <td>6.960253</td>\n",
       "      <td>6.991211</td>\n",
       "      <td>6.953835</td>\n",
       "      <td>6.831518</td>\n",
       "      <td>7.015863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>neuronal nitric-oxide synthase isoform mu</th>\n",
       "      <td>6.429438</td>\n",
       "      <td>6.674736</td>\n",
       "      <td>6.401588</td>\n",
       "      <td>6.363483</td>\n",
       "      <td>6.342921</td>\n",
       "      <td>6.566346</td>\n",
       "      <td>6.379514</td>\n",
       "      <td>6.349456</td>\n",
       "      <td>6.532331</td>\n",
       "      <td>6.577870</td>\n",
       "      <td>...</td>\n",
       "      <td>6.470454</td>\n",
       "      <td>6.517419</td>\n",
       "      <td>6.548269</td>\n",
       "      <td>6.485421</td>\n",
       "      <td>6.447770</td>\n",
       "      <td>6.461752</td>\n",
       "      <td>6.479435</td>\n",
       "      <td>6.402252</td>\n",
       "      <td>6.610253</td>\n",
       "      <td>6.398350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p</th>\n",
       "      <td>6.541973</td>\n",
       "      <td>6.359801</td>\n",
       "      <td>6.517618</td>\n",
       "      <td>6.426932</td>\n",
       "      <td>6.535042</td>\n",
       "      <td>6.479159</td>\n",
       "      <td>6.212613</td>\n",
       "      <td>6.386862</td>\n",
       "      <td>6.508521</td>\n",
       "      <td>6.347702</td>\n",
       "      <td>...</td>\n",
       "      <td>6.104731</td>\n",
       "      <td>6.359801</td>\n",
       "      <td>6.466790</td>\n",
       "      <td>6.642090</td>\n",
       "      <td>6.369591</td>\n",
       "      <td>6.415152</td>\n",
       "      <td>6.336126</td>\n",
       "      <td>6.395560</td>\n",
       "      <td>6.400445</td>\n",
       "      <td>6.347738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scavenger receptor</th>\n",
       "      <td>7.106817</td>\n",
       "      <td>7.059044</td>\n",
       "      <td>6.899810</td>\n",
       "      <td>6.699562</td>\n",
       "      <td>7.049678</td>\n",
       "      <td>7.526465</td>\n",
       "      <td>7.402048</td>\n",
       "      <td>7.746785</td>\n",
       "      <td>6.798596</td>\n",
       "      <td>8.034765</td>\n",
       "      <td>...</td>\n",
       "      <td>6.715678</td>\n",
       "      <td>7.140271</td>\n",
       "      <td>6.874077</td>\n",
       "      <td>7.132313</td>\n",
       "      <td>7.834666</td>\n",
       "      <td>7.702109</td>\n",
       "      <td>6.960724</td>\n",
       "      <td>6.652632</td>\n",
       "      <td>7.252858</td>\n",
       "      <td>6.451716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sty</th>\n",
       "      <td>8.787945</td>\n",
       "      <td>8.523423</td>\n",
       "      <td>7.741574</td>\n",
       "      <td>9.063488</td>\n",
       "      <td>8.008797</td>\n",
       "      <td>9.646043</td>\n",
       "      <td>9.472719</td>\n",
       "      <td>8.881412</td>\n",
       "      <td>9.161894</td>\n",
       "      <td>9.242461</td>\n",
       "      <td>...</td>\n",
       "      <td>9.029700</td>\n",
       "      <td>9.350928</td>\n",
       "      <td>9.545547</td>\n",
       "      <td>9.239431</td>\n",
       "      <td>9.492746</td>\n",
       "      <td>10.106197</td>\n",
       "      <td>9.241477</td>\n",
       "      <td>8.622802</td>\n",
       "      <td>9.116669</td>\n",
       "      <td>9.550628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tea</th>\n",
       "      <td>6.494957</td>\n",
       "      <td>6.497032</td>\n",
       "      <td>6.438302</td>\n",
       "      <td>6.627027</td>\n",
       "      <td>6.473407</td>\n",
       "      <td>6.454031</td>\n",
       "      <td>6.418251</td>\n",
       "      <td>6.405716</td>\n",
       "      <td>6.471698</td>\n",
       "      <td>6.501229</td>\n",
       "      <td>...</td>\n",
       "      <td>6.535940</td>\n",
       "      <td>6.373899</td>\n",
       "      <td>6.444591</td>\n",
       "      <td>6.471909</td>\n",
       "      <td>6.478121</td>\n",
       "      <td>6.372046</td>\n",
       "      <td>6.570508</td>\n",
       "      <td>6.388176</td>\n",
       "      <td>6.506383</td>\n",
       "      <td>6.417805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30122 rows × 21 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:41.772528Z",
     "start_time": "2024-09-20T14:41:39.149327Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Get the liver alphabet\n",
    "liver_df = parse_gse(gse_liver)\n",
    "# Merge the metadata with the alphabet\n",
    "liver_df = merge(liver_df, metadata_liver, \"GSM_name\", \"GSM name\")\n",
    "# Merge the gene identifiers with the alphabet\n",
    "liver_df = merge(liver_df, gene_identifiers_df_liver, \"ID_REF\", \"ID\")\n",
    "# Drop technical replicates\n",
    "liver_df = filter_samples(liver_df, \"Liver\")\n",
    "# Reshape the alphabet\n",
    "liver_df = pivot_strains(liver_df, \"GENE_NAME\")\n",
    "# Calculate the mean across sex for each strain\n",
    "liver_df = mean_batches(liver_df, \"Liver\")\n",
    "liver_df"
   ],
   "id": "6f014763a05d5dab",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "strain                                                B6D2F1    BXD1  BXD11  \\\n",
       "GENE_NAME                                                                     \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  1.210833  1.7480  2.243   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.598000  1.7625  1.905   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.421167  0.7030  0.508   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 4 ... -2.789833 -2.7585 -2.945   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -0.886667 -0.6520 -1.477   \n",
       "...                                                      ...     ...    ...   \n",
       "zona pellucida like domain containing 1            -0.206167 -0.1685 -0.153   \n",
       "zonadhesin                                          0.279500  0.2415  0.438   \n",
       "zyg-11 homolog A (C. elegans)                       0.458167  0.4280  0.728   \n",
       "zygote arrest 1                                     0.038667 -0.1490 -0.001   \n",
       "zyxin                                              -0.307333 -0.5450 -0.866   \n",
       "\n",
       "strain                                              BXD11TY   BXD12   BXD13  \\\n",
       "GENE_NAME                                                                     \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...    1.113  1.2085  1.9485   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...    1.981  1.5790  1.9415   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3        0.040  0.5605  0.3715   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 4 ...   -2.986 -2.7850 -3.0230   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ...   -1.480 -0.5300 -1.1370   \n",
       "...                                                     ...     ...     ...   \n",
       "zona pellucida like domain containing 1              -0.094 -0.1950 -1.0575   \n",
       "zonadhesin                                            0.350  0.2925  0.2875   \n",
       "zyg-11 homolog A (C. elegans)                         0.773  0.3185  0.2415   \n",
       "zygote arrest 1                                       0.417  0.2700 -0.1700   \n",
       "zyxin                                                -0.487 -0.5945 -0.7400   \n",
       "\n",
       "strain                                               BXD14   BXD15   BXD16  \\\n",
       "GENE_NAME                                                                    \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  0.9215  1.0310  1.1545   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.3830  1.8660  2.0565   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.2740  0.6665  0.4685   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 4 ... -2.6615 -2.7080 -3.2560   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -0.8910 -0.2990 -1.7375   \n",
       "...                                                    ...     ...     ...   \n",
       "zona pellucida like domain containing 1            -0.0735 -0.2225 -0.0810   \n",
       "zonadhesin                                          0.3600  0.2525  0.3925   \n",
       "zyg-11 homolog A (C. elegans)                       0.6095  0.0635  0.6660   \n",
       "zygote arrest 1                                    -0.1110 -0.0745  0.4875   \n",
       "zyxin                                              -0.4270 -0.3235 -0.6515   \n",
       "\n",
       "strain                                               BXD19  ...   BXD69  \\\n",
       "GENE_NAME                                                   ...           \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  1.5565  ...  1.2765   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  2.3185  ...  1.9495   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.3900  ...  0.4090   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 4 ... -3.1915  ... -2.9405   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -1.6430  ... -1.6545   \n",
       "...                                                    ...  ...     ...   \n",
       "zona pellucida like domain containing 1            -0.1225  ...  0.0140   \n",
       "zonadhesin                                          0.2340  ...  0.0630   \n",
       "zyg-11 homolog A (C. elegans)                       0.4840  ...  0.5350   \n",
       "zygote arrest 1                                    -0.0955  ... -0.0925   \n",
       "zyxin                                              -0.4150  ... -0.5060   \n",
       "\n",
       "strain                                               BXD73   BXD77    BXD8  \\\n",
       "GENE_NAME                                                                    \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  0.7425  1.4360  1.7245   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.4490  1.8260  2.0670   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.2680  0.4305  0.6770   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 4 ... -2.8110 -3.0055 -2.9290   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -1.7400 -0.6750 -0.4900   \n",
       "...                                                    ...     ...     ...   \n",
       "zona pellucida like domain containing 1            -0.1100 -0.3550  0.0940   \n",
       "zonadhesin                                          0.3440  0.0840  0.2090   \n",
       "zyg-11 homolog A (C. elegans)                       0.2815  0.7390  0.6840   \n",
       "zygote arrest 1                                     0.3760  0.3775  0.2995   \n",
       "zyxin                                              -0.3360 -0.3805 -0.6185   \n",
       "\n",
       "strain                                               BXD85   BXD86    BXD9  \\\n",
       "GENE_NAME                                                                    \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  1.3890  1.4395  1.1540   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.7725  1.8180  1.2430   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.3760  0.4680  0.5405   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 4 ... -3.1155 -2.9195 -2.7315   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -1.8610 -0.6505 -0.5690   \n",
       "...                                                    ...     ...     ...   \n",
       "zona pellucida like domain containing 1            -0.1725 -0.3360  0.0930   \n",
       "zonadhesin                                          0.2685  0.1960  0.7185   \n",
       "zyg-11 homolog A (C. elegans)                       0.3355  0.6170  0.5690   \n",
       "zygote arrest 1                                    -0.2295  0.5980 -0.1330   \n",
       "zyxin                                              -0.4655 -0.1315 -0.5515   \n",
       "\n",
       "strain                                               BXD92  C57BL6J   DBA2J  \n",
       "GENE_NAME                                                                    \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  1.3800  1.12200  1.3166  \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.7840  1.75100  1.8952  \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.2110  0.53600  0.6947  \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 4 ... -2.9565 -2.92500 -2.6595  \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -0.6770 -0.61000 -1.3286  \n",
       "...                                                    ...      ...     ...  \n",
       "zona pellucida like domain containing 1            -0.1780 -0.03725 -0.0173  \n",
       "zonadhesin                                          0.0780  0.40250  0.3905  \n",
       "zyg-11 homolog A (C. elegans)                       0.6635  0.44575  0.2731  \n",
       "zygote arrest 1                                    -0.1920 -0.19500  0.1497  \n",
       "zyxin                                              -0.3000 -0.45625 -0.4278  \n",
       "\n",
       "[17917 rows x 44 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strain</th>\n",
       "      <th>B6D2F1</th>\n",
       "      <th>BXD1</th>\n",
       "      <th>BXD11</th>\n",
       "      <th>BXD11TY</th>\n",
       "      <th>BXD12</th>\n",
       "      <th>BXD13</th>\n",
       "      <th>BXD14</th>\n",
       "      <th>BXD15</th>\n",
       "      <th>BXD16</th>\n",
       "      <th>BXD19</th>\n",
       "      <th>...</th>\n",
       "      <th>BXD69</th>\n",
       "      <th>BXD73</th>\n",
       "      <th>BXD77</th>\n",
       "      <th>BXD8</th>\n",
       "      <th>BXD85</th>\n",
       "      <th>BXD86</th>\n",
       "      <th>BXD9</th>\n",
       "      <th>BXD92</th>\n",
       "      <th>C57BL6J</th>\n",
       "      <th>DBA2J</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1-acylglycerol-3-phosphate O-acyltransferase 1 (lysophosphatidic acid acyltransferase, alpha)\"</th>\n",
       "      <td>1.210833</td>\n",
       "      <td>1.7480</td>\n",
       "      <td>2.243</td>\n",
       "      <td>1.113</td>\n",
       "      <td>1.2085</td>\n",
       "      <td>1.9485</td>\n",
       "      <td>0.9215</td>\n",
       "      <td>1.0310</td>\n",
       "      <td>1.1545</td>\n",
       "      <td>1.5565</td>\n",
       "      <td>...</td>\n",
       "      <td>1.2765</td>\n",
       "      <td>0.7425</td>\n",
       "      <td>1.4360</td>\n",
       "      <td>1.7245</td>\n",
       "      <td>1.3890</td>\n",
       "      <td>1.4395</td>\n",
       "      <td>1.1540</td>\n",
       "      <td>1.3800</td>\n",
       "      <td>1.12200</td>\n",
       "      <td>1.3166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1-acylglycerol-3-phosphate O-acyltransferase 2 (lysophosphatidic acid acyltransferase, beta)</th>\n",
       "      <td>1.598000</td>\n",
       "      <td>1.7625</td>\n",
       "      <td>1.905</td>\n",
       "      <td>1.981</td>\n",
       "      <td>1.5790</td>\n",
       "      <td>1.9415</td>\n",
       "      <td>1.3830</td>\n",
       "      <td>1.8660</td>\n",
       "      <td>2.0565</td>\n",
       "      <td>2.3185</td>\n",
       "      <td>...</td>\n",
       "      <td>1.9495</td>\n",
       "      <td>1.4490</td>\n",
       "      <td>1.8260</td>\n",
       "      <td>2.0670</td>\n",
       "      <td>1.7725</td>\n",
       "      <td>1.8180</td>\n",
       "      <td>1.2430</td>\n",
       "      <td>1.7840</td>\n",
       "      <td>1.75100</td>\n",
       "      <td>1.8952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1-acylglycerol-3-phosphate O-acyltransferase 3</th>\n",
       "      <td>0.421167</td>\n",
       "      <td>0.7030</td>\n",
       "      <td>0.508</td>\n",
       "      <td>0.040</td>\n",
       "      <td>0.5605</td>\n",
       "      <td>0.3715</td>\n",
       "      <td>0.2740</td>\n",
       "      <td>0.6665</td>\n",
       "      <td>0.4685</td>\n",
       "      <td>0.3900</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4090</td>\n",
       "      <td>0.2680</td>\n",
       "      <td>0.4305</td>\n",
       "      <td>0.6770</td>\n",
       "      <td>0.3760</td>\n",
       "      <td>0.4680</td>\n",
       "      <td>0.5405</td>\n",
       "      <td>0.2110</td>\n",
       "      <td>0.53600</td>\n",
       "      <td>0.6947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1-acylglycerol-3-phosphate O-acyltransferase 4 (lysophosphatidic acid acyltransferase, delta)\"</th>\n",
       "      <td>-2.789833</td>\n",
       "      <td>-2.7585</td>\n",
       "      <td>-2.945</td>\n",
       "      <td>-2.986</td>\n",
       "      <td>-2.7850</td>\n",
       "      <td>-3.0230</td>\n",
       "      <td>-2.6615</td>\n",
       "      <td>-2.7080</td>\n",
       "      <td>-3.2560</td>\n",
       "      <td>-3.1915</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.9405</td>\n",
       "      <td>-2.8110</td>\n",
       "      <td>-3.0055</td>\n",
       "      <td>-2.9290</td>\n",
       "      <td>-3.1155</td>\n",
       "      <td>-2.9195</td>\n",
       "      <td>-2.7315</td>\n",
       "      <td>-2.9565</td>\n",
       "      <td>-2.92500</td>\n",
       "      <td>-2.6595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1-acylglycerol-3-phosphate O-acyltransferase 5 (lysophosphatidic acid acyltransferase, epsilon)</th>\n",
       "      <td>-0.886667</td>\n",
       "      <td>-0.6520</td>\n",
       "      <td>-1.477</td>\n",
       "      <td>-1.480</td>\n",
       "      <td>-0.5300</td>\n",
       "      <td>-1.1370</td>\n",
       "      <td>-0.8910</td>\n",
       "      <td>-0.2990</td>\n",
       "      <td>-1.7375</td>\n",
       "      <td>-1.6430</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.6545</td>\n",
       "      <td>-1.7400</td>\n",
       "      <td>-0.6750</td>\n",
       "      <td>-0.4900</td>\n",
       "      <td>-1.8610</td>\n",
       "      <td>-0.6505</td>\n",
       "      <td>-0.5690</td>\n",
       "      <td>-0.6770</td>\n",
       "      <td>-0.61000</td>\n",
       "      <td>-1.3286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zona pellucida like domain containing 1</th>\n",
       "      <td>-0.206167</td>\n",
       "      <td>-0.1685</td>\n",
       "      <td>-0.153</td>\n",
       "      <td>-0.094</td>\n",
       "      <td>-0.1950</td>\n",
       "      <td>-1.0575</td>\n",
       "      <td>-0.0735</td>\n",
       "      <td>-0.2225</td>\n",
       "      <td>-0.0810</td>\n",
       "      <td>-0.1225</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0140</td>\n",
       "      <td>-0.1100</td>\n",
       "      <td>-0.3550</td>\n",
       "      <td>0.0940</td>\n",
       "      <td>-0.1725</td>\n",
       "      <td>-0.3360</td>\n",
       "      <td>0.0930</td>\n",
       "      <td>-0.1780</td>\n",
       "      <td>-0.03725</td>\n",
       "      <td>-0.0173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zonadhesin</th>\n",
       "      <td>0.279500</td>\n",
       "      <td>0.2415</td>\n",
       "      <td>0.438</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.2925</td>\n",
       "      <td>0.2875</td>\n",
       "      <td>0.3600</td>\n",
       "      <td>0.2525</td>\n",
       "      <td>0.3925</td>\n",
       "      <td>0.2340</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0630</td>\n",
       "      <td>0.3440</td>\n",
       "      <td>0.0840</td>\n",
       "      <td>0.2090</td>\n",
       "      <td>0.2685</td>\n",
       "      <td>0.1960</td>\n",
       "      <td>0.7185</td>\n",
       "      <td>0.0780</td>\n",
       "      <td>0.40250</td>\n",
       "      <td>0.3905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zyg-11 homolog A (C. elegans)</th>\n",
       "      <td>0.458167</td>\n",
       "      <td>0.4280</td>\n",
       "      <td>0.728</td>\n",
       "      <td>0.773</td>\n",
       "      <td>0.3185</td>\n",
       "      <td>0.2415</td>\n",
       "      <td>0.6095</td>\n",
       "      <td>0.0635</td>\n",
       "      <td>0.6660</td>\n",
       "      <td>0.4840</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5350</td>\n",
       "      <td>0.2815</td>\n",
       "      <td>0.7390</td>\n",
       "      <td>0.6840</td>\n",
       "      <td>0.3355</td>\n",
       "      <td>0.6170</td>\n",
       "      <td>0.5690</td>\n",
       "      <td>0.6635</td>\n",
       "      <td>0.44575</td>\n",
       "      <td>0.2731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zygote arrest 1</th>\n",
       "      <td>0.038667</td>\n",
       "      <td>-0.1490</td>\n",
       "      <td>-0.001</td>\n",
       "      <td>0.417</td>\n",
       "      <td>0.2700</td>\n",
       "      <td>-0.1700</td>\n",
       "      <td>-0.1110</td>\n",
       "      <td>-0.0745</td>\n",
       "      <td>0.4875</td>\n",
       "      <td>-0.0955</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.0925</td>\n",
       "      <td>0.3760</td>\n",
       "      <td>0.3775</td>\n",
       "      <td>0.2995</td>\n",
       "      <td>-0.2295</td>\n",
       "      <td>0.5980</td>\n",
       "      <td>-0.1330</td>\n",
       "      <td>-0.1920</td>\n",
       "      <td>-0.19500</td>\n",
       "      <td>0.1497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zyxin</th>\n",
       "      <td>-0.307333</td>\n",
       "      <td>-0.5450</td>\n",
       "      <td>-0.866</td>\n",
       "      <td>-0.487</td>\n",
       "      <td>-0.5945</td>\n",
       "      <td>-0.7400</td>\n",
       "      <td>-0.4270</td>\n",
       "      <td>-0.3235</td>\n",
       "      <td>-0.6515</td>\n",
       "      <td>-0.4150</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5060</td>\n",
       "      <td>-0.3360</td>\n",
       "      <td>-0.3805</td>\n",
       "      <td>-0.6185</td>\n",
       "      <td>-0.4655</td>\n",
       "      <td>-0.1315</td>\n",
       "      <td>-0.5515</td>\n",
       "      <td>-0.3000</td>\n",
       "      <td>-0.45625</td>\n",
       "      <td>-0.4278</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17917 rows × 44 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 13
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:41.781916Z",
     "start_time": "2024-09-20T14:41:41.773370Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Filter the relevant strains to include only those present in liver_df\n",
    "filtered_strains_liver = [col for col in relevant_strains if col in liver_df.columns]\n",
    "# Select the filtered columns from liver_df\n",
    "liver_df = liver_df.loc[:, filtered_strains_liver]\n",
    "# Get only the strains present in liver_df from the genotype alphabet\n",
    "# Filter the genotype alphabet to include only the relevant strains\n",
    "representative_snps_liver = genotype_df[filtered_strains_liver]\n",
    "# Now look for snps that have exactly same values for all the strains\n",
    "duplicates = representative_snps_liver[representative_snps_liver.duplicated()]\n",
    "print(f\"The number of duplicate SNPs is: {len(duplicates)}\")\n",
    "# Remove the duplicate SNPs\n",
    "representative_snps_liver = representative_snps_liver.drop_duplicates()"
   ],
   "id": "aa3bdd4e1916eb78",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of duplicate SNPs is: 2424\n"
     ]
    }
   ],
   "execution_count": 14
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:41.809525Z",
     "start_time": "2024-09-20T14:41:41.782813Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Filtering for liver alphabet\n",
    "liver_df_filtered = liver_df.dropna()\n",
    "row_variance = liver_df_filtered.var(axis=1)\n",
    "average_variance = row_variance.mean()\n",
    "print(f\"The average variance is: {round(average_variance, 3)}\")\n",
    "liver_df_filtered = liver_df_filtered[row_variance > average_variance/10]\n",
    "row_max = liver_df_filtered.max(axis=1)\n",
    "average_max = row_max.mean()\n",
    "print(f\"The average maximum expression value is: {round(average_max, 3)}\")\n",
    "liver_df_filtered = liver_df_filtered[row_max > average_max-2]\n",
    "print(f\"The number of genes before filtering is: {len(liver_df)} \\nThe number of genes after filtering is: {len(liver_df_filtered)}\")\n",
    "liver_df_filtered"
   ],
   "id": "367b063e2190a0a4",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The average variance is: 0.055\n",
      "The average maximum expression value is: 0.275\n",
      "The number of genes before filtering is: 17917 \n",
      "The number of genes after filtering is: 16864\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "strain                                                BXD1    BXD2    BXD5  \\\n",
       "GENE_NAME                                                                    \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  1.7480  1.1440  0.8800   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.7625  1.8615  1.7190   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.7030  0.2645  0.6325   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -0.6520 -0.1725 -1.1295   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 6 ...  0.8450  0.7320  1.1730   \n",
       "...                                                    ...     ...     ...   \n",
       "zona pellucida like domain containing 1            -0.1685 -0.1325 -0.4075   \n",
       "zonadhesin                                          0.2415  0.3035  0.2125   \n",
       "zyg-11 homolog A (C. elegans)                       0.4280  0.8640  0.2355   \n",
       "zygote arrest 1                                    -0.1490  0.3805  0.4180   \n",
       "zyxin                                              -0.5450 -0.6260 -0.2665   \n",
       "\n",
       "strain                                                BXD6    BXD8    BXD9  \\\n",
       "GENE_NAME                                                                    \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  1.2030  1.7245  1.1540   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.6200  2.0670  1.2430   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.3815  0.6770  0.5405   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -0.4600 -0.4900 -0.5690   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 6 ...  0.9065  1.1170  1.6770   \n",
       "...                                                    ...     ...     ...   \n",
       "zona pellucida like domain containing 1            -0.3370  0.0940  0.0930   \n",
       "zonadhesin                                          0.3725  0.2090  0.7185   \n",
       "zyg-11 homolog A (C. elegans)                       0.3880  0.6840  0.5690   \n",
       "zygote arrest 1                                     0.4005  0.2995 -0.1330   \n",
       "zyxin                                              -0.2855 -0.6185 -0.5515   \n",
       "\n",
       "strain                                              BXD11   BXD12   BXD13  \\\n",
       "GENE_NAME                                                                   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  2.243  1.2085  1.9485   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.905  1.5790  1.9415   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.508  0.5605  0.3715   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -1.477 -0.5300 -1.1370   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 6 ...  1.267  1.2195  1.4020   \n",
       "...                                                   ...     ...     ...   \n",
       "zona pellucida like domain containing 1            -0.153 -0.1950 -1.0575   \n",
       "zonadhesin                                          0.438  0.2925  0.2875   \n",
       "zyg-11 homolog A (C. elegans)                       0.728  0.3185  0.2415   \n",
       "zygote arrest 1                                    -0.001  0.2700 -0.1700   \n",
       "zyxin                                              -0.866 -0.5945 -0.7400   \n",
       "\n",
       "strain                                               BXD14  ...   BXD48  \\\n",
       "GENE_NAME                                                   ...           \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  0.9215  ...  0.8665   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.3830  ...  2.1070   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.2740  ...  0.2860   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -0.8910  ... -0.5345   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 6 ...  0.9095  ...  0.8305   \n",
       "...                                                    ...  ...     ...   \n",
       "zona pellucida like domain containing 1            -0.0735  ... -0.1560   \n",
       "zonadhesin                                          0.3600  ...  0.3075   \n",
       "zyg-11 homolog A (C. elegans)                       0.6095  ...  0.4615   \n",
       "zygote arrest 1                                    -0.1110  ...  0.5390   \n",
       "zyxin                                              -0.4270  ... -0.6000   \n",
       "\n",
       "strain                                               BXD51   BXD60   BXD62  \\\n",
       "GENE_NAME                                                                    \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  0.9135  1.0820  0.9410   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.9635  1.5355  1.7440   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.3145  0.5700  0.5715   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -0.7830 -1.2535 -0.7695   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 6 ...  1.1930  0.9895  1.6265   \n",
       "...                                                    ...     ...     ...   \n",
       "zona pellucida like domain containing 1            -0.1410 -0.0850 -0.1955   \n",
       "zonadhesin                                          0.2205  0.2730  0.2175   \n",
       "zyg-11 homolog A (C. elegans)                       0.5400  0.3875  0.5035   \n",
       "zygote arrest 1                                     0.4385  0.2885  0.3260   \n",
       "zyxin                                              -0.1405 -0.3165 -0.2665   \n",
       "\n",
       "strain                                               BXD69   BXD73   BXD77  \\\n",
       "GENE_NAME                                                                    \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  1.2765  0.7425  1.4360   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.9495  1.4490  1.8260   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.4090  0.2680  0.4305   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -1.6545 -1.7400 -0.6750   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 6 ...  1.3075  1.7400  0.9195   \n",
       "...                                                    ...     ...     ...   \n",
       "zona pellucida like domain containing 1             0.0140 -0.1100 -0.3550   \n",
       "zonadhesin                                          0.0630  0.3440  0.0840   \n",
       "zyg-11 homolog A (C. elegans)                       0.5350  0.2815  0.7390   \n",
       "zygote arrest 1                                    -0.0925  0.3760  0.3775   \n",
       "zyxin                                              -0.5060 -0.3360 -0.3805   \n",
       "\n",
       "strain                                               BXD85   BXD86   BXD92  \n",
       "GENE_NAME                                                                   \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 1 ...  1.3890  1.4395  1.3800  \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 2 ...  1.7725  1.8180  1.7840  \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 3      0.3760  0.4680  0.2110  \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 5 ... -1.8610 -0.6505 -0.6770  \n",
       "1-acylglycerol-3-phosphate O-acyltransferase 6 ...  1.2785  0.9790  1.3175  \n",
       "...                                                    ...     ...     ...  \n",
       "zona pellucida like domain containing 1            -0.1725 -0.3360 -0.1780  \n",
       "zonadhesin                                          0.2685  0.1960  0.0780  \n",
       "zyg-11 homolog A (C. elegans)                       0.3355  0.6170  0.6635  \n",
       "zygote arrest 1                                    -0.2295  0.5980 -0.1920  \n",
       "zyxin                                              -0.4655 -0.1315 -0.3000  \n",
       "\n",
       "[16864 rows x 40 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>strain</th>\n",
       "      <th>BXD1</th>\n",
       "      <th>BXD2</th>\n",
       "      <th>BXD5</th>\n",
       "      <th>BXD6</th>\n",
       "      <th>BXD8</th>\n",
       "      <th>BXD9</th>\n",
       "      <th>BXD11</th>\n",
       "      <th>BXD12</th>\n",
       "      <th>BXD13</th>\n",
       "      <th>BXD14</th>\n",
       "      <th>...</th>\n",
       "      <th>BXD48</th>\n",
       "      <th>BXD51</th>\n",
       "      <th>BXD60</th>\n",
       "      <th>BXD62</th>\n",
       "      <th>BXD69</th>\n",
       "      <th>BXD73</th>\n",
       "      <th>BXD77</th>\n",
       "      <th>BXD85</th>\n",
       "      <th>BXD86</th>\n",
       "      <th>BXD92</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1-acylglycerol-3-phosphate O-acyltransferase 1 (lysophosphatidic acid acyltransferase, alpha)\"</th>\n",
       "      <td>1.7480</td>\n",
       "      <td>1.1440</td>\n",
       "      <td>0.8800</td>\n",
       "      <td>1.2030</td>\n",
       "      <td>1.7245</td>\n",
       "      <td>1.1540</td>\n",
       "      <td>2.243</td>\n",
       "      <td>1.2085</td>\n",
       "      <td>1.9485</td>\n",
       "      <td>0.9215</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8665</td>\n",
       "      <td>0.9135</td>\n",
       "      <td>1.0820</td>\n",
       "      <td>0.9410</td>\n",
       "      <td>1.2765</td>\n",
       "      <td>0.7425</td>\n",
       "      <td>1.4360</td>\n",
       "      <td>1.3890</td>\n",
       "      <td>1.4395</td>\n",
       "      <td>1.3800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1-acylglycerol-3-phosphate O-acyltransferase 2 (lysophosphatidic acid acyltransferase, beta)</th>\n",
       "      <td>1.7625</td>\n",
       "      <td>1.8615</td>\n",
       "      <td>1.7190</td>\n",
       "      <td>1.6200</td>\n",
       "      <td>2.0670</td>\n",
       "      <td>1.2430</td>\n",
       "      <td>1.905</td>\n",
       "      <td>1.5790</td>\n",
       "      <td>1.9415</td>\n",
       "      <td>1.3830</td>\n",
       "      <td>...</td>\n",
       "      <td>2.1070</td>\n",
       "      <td>1.9635</td>\n",
       "      <td>1.5355</td>\n",
       "      <td>1.7440</td>\n",
       "      <td>1.9495</td>\n",
       "      <td>1.4490</td>\n",
       "      <td>1.8260</td>\n",
       "      <td>1.7725</td>\n",
       "      <td>1.8180</td>\n",
       "      <td>1.7840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1-acylglycerol-3-phosphate O-acyltransferase 3</th>\n",
       "      <td>0.7030</td>\n",
       "      <td>0.2645</td>\n",
       "      <td>0.6325</td>\n",
       "      <td>0.3815</td>\n",
       "      <td>0.6770</td>\n",
       "      <td>0.5405</td>\n",
       "      <td>0.508</td>\n",
       "      <td>0.5605</td>\n",
       "      <td>0.3715</td>\n",
       "      <td>0.2740</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2860</td>\n",
       "      <td>0.3145</td>\n",
       "      <td>0.5700</td>\n",
       "      <td>0.5715</td>\n",
       "      <td>0.4090</td>\n",
       "      <td>0.2680</td>\n",
       "      <td>0.4305</td>\n",
       "      <td>0.3760</td>\n",
       "      <td>0.4680</td>\n",
       "      <td>0.2110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1-acylglycerol-3-phosphate O-acyltransferase 5 (lysophosphatidic acid acyltransferase, epsilon)</th>\n",
       "      <td>-0.6520</td>\n",
       "      <td>-0.1725</td>\n",
       "      <td>-1.1295</td>\n",
       "      <td>-0.4600</td>\n",
       "      <td>-0.4900</td>\n",
       "      <td>-0.5690</td>\n",
       "      <td>-1.477</td>\n",
       "      <td>-0.5300</td>\n",
       "      <td>-1.1370</td>\n",
       "      <td>-0.8910</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5345</td>\n",
       "      <td>-0.7830</td>\n",
       "      <td>-1.2535</td>\n",
       "      <td>-0.7695</td>\n",
       "      <td>-1.6545</td>\n",
       "      <td>-1.7400</td>\n",
       "      <td>-0.6750</td>\n",
       "      <td>-1.8610</td>\n",
       "      <td>-0.6505</td>\n",
       "      <td>-0.6770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1-acylglycerol-3-phosphate O-acyltransferase 6 (lysophosphatidic acid acyltransferase, zeta)</th>\n",
       "      <td>0.8450</td>\n",
       "      <td>0.7320</td>\n",
       "      <td>1.1730</td>\n",
       "      <td>0.9065</td>\n",
       "      <td>1.1170</td>\n",
       "      <td>1.6770</td>\n",
       "      <td>1.267</td>\n",
       "      <td>1.2195</td>\n",
       "      <td>1.4020</td>\n",
       "      <td>0.9095</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8305</td>\n",
       "      <td>1.1930</td>\n",
       "      <td>0.9895</td>\n",
       "      <td>1.6265</td>\n",
       "      <td>1.3075</td>\n",
       "      <td>1.7400</td>\n",
       "      <td>0.9195</td>\n",
       "      <td>1.2785</td>\n",
       "      <td>0.9790</td>\n",
       "      <td>1.3175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zona pellucida like domain containing 1</th>\n",
       "      <td>-0.1685</td>\n",
       "      <td>-0.1325</td>\n",
       "      <td>-0.4075</td>\n",
       "      <td>-0.3370</td>\n",
       "      <td>0.0940</td>\n",
       "      <td>0.0930</td>\n",
       "      <td>-0.153</td>\n",
       "      <td>-0.1950</td>\n",
       "      <td>-1.0575</td>\n",
       "      <td>-0.0735</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.1560</td>\n",
       "      <td>-0.1410</td>\n",
       "      <td>-0.0850</td>\n",
       "      <td>-0.1955</td>\n",
       "      <td>0.0140</td>\n",
       "      <td>-0.1100</td>\n",
       "      <td>-0.3550</td>\n",
       "      <td>-0.1725</td>\n",
       "      <td>-0.3360</td>\n",
       "      <td>-0.1780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zonadhesin</th>\n",
       "      <td>0.2415</td>\n",
       "      <td>0.3035</td>\n",
       "      <td>0.2125</td>\n",
       "      <td>0.3725</td>\n",
       "      <td>0.2090</td>\n",
       "      <td>0.7185</td>\n",
       "      <td>0.438</td>\n",
       "      <td>0.2925</td>\n",
       "      <td>0.2875</td>\n",
       "      <td>0.3600</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3075</td>\n",
       "      <td>0.2205</td>\n",
       "      <td>0.2730</td>\n",
       "      <td>0.2175</td>\n",
       "      <td>0.0630</td>\n",
       "      <td>0.3440</td>\n",
       "      <td>0.0840</td>\n",
       "      <td>0.2685</td>\n",
       "      <td>0.1960</td>\n",
       "      <td>0.0780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zyg-11 homolog A (C. elegans)</th>\n",
       "      <td>0.4280</td>\n",
       "      <td>0.8640</td>\n",
       "      <td>0.2355</td>\n",
       "      <td>0.3880</td>\n",
       "      <td>0.6840</td>\n",
       "      <td>0.5690</td>\n",
       "      <td>0.728</td>\n",
       "      <td>0.3185</td>\n",
       "      <td>0.2415</td>\n",
       "      <td>0.6095</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4615</td>\n",
       "      <td>0.5400</td>\n",
       "      <td>0.3875</td>\n",
       "      <td>0.5035</td>\n",
       "      <td>0.5350</td>\n",
       "      <td>0.2815</td>\n",
       "      <td>0.7390</td>\n",
       "      <td>0.3355</td>\n",
       "      <td>0.6170</td>\n",
       "      <td>0.6635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zygote arrest 1</th>\n",
       "      <td>-0.1490</td>\n",
       "      <td>0.3805</td>\n",
       "      <td>0.4180</td>\n",
       "      <td>0.4005</td>\n",
       "      <td>0.2995</td>\n",
       "      <td>-0.1330</td>\n",
       "      <td>-0.001</td>\n",
       "      <td>0.2700</td>\n",
       "      <td>-0.1700</td>\n",
       "      <td>-0.1110</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5390</td>\n",
       "      <td>0.4385</td>\n",
       "      <td>0.2885</td>\n",
       "      <td>0.3260</td>\n",
       "      <td>-0.0925</td>\n",
       "      <td>0.3760</td>\n",
       "      <td>0.3775</td>\n",
       "      <td>-0.2295</td>\n",
       "      <td>0.5980</td>\n",
       "      <td>-0.1920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zyxin</th>\n",
       "      <td>-0.5450</td>\n",
       "      <td>-0.6260</td>\n",
       "      <td>-0.2665</td>\n",
       "      <td>-0.2855</td>\n",
       "      <td>-0.6185</td>\n",
       "      <td>-0.5515</td>\n",
       "      <td>-0.866</td>\n",
       "      <td>-0.5945</td>\n",
       "      <td>-0.7400</td>\n",
       "      <td>-0.4270</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.6000</td>\n",
       "      <td>-0.1405</td>\n",
       "      <td>-0.3165</td>\n",
       "      <td>-0.2665</td>\n",
       "      <td>-0.5060</td>\n",
       "      <td>-0.3360</td>\n",
       "      <td>-0.3805</td>\n",
       "      <td>-0.4655</td>\n",
       "      <td>-0.1315</td>\n",
       "      <td>-0.3000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16864 rows × 40 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 15
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:41:41.843769Z",
     "start_time": "2024-09-20T14:41:41.810452Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Function for the analysis \n",
    "def association_test(df, snps):\n",
    "    eQTLs = []\n",
    "    for gene in df.index:\n",
    "        # Run the regression\n",
    "        results = f_regression(snps.T, df.loc[gene].T)\n",
    "        # Get the p-values\n",
    "        p_values = results[1]\n",
    "        eQTLs.append({\"gene\": gene, \"p_values\": p_values})\n",
    "    eQTLs_df = pd.DataFrame(eQTLs)\n",
    "    return eQTLs_df\n",
    "\n",
    "def correct_results(df1, df2):\n",
    "    # Merge the p-values columns for FDR correction\n",
    "    p_merged = pd.concat([df1['p_values'], df2['p_values']], axis=0)\n",
    "    p_values = list(chain.from_iterable(p_merged))\n",
    "    \n",
    "    # Correct the p-values for multiple testing\n",
    "    corrected_p_values = fdrcorrection(p_values)[1]\n",
    "    \n",
    "    # Initialize new columns in both dataframes\n",
    "    df1[\"corrected_p_values\"] = [None] * len(df1)\n",
    "    df1[\"significant_snps\"] = [None] * len(df1)\n",
    "    df2[\"corrected_p_values\"] = [None] * len(df2)\n",
    "    df2[\"significant_snps\"] = [None] * len(df2)\n",
    "    \n",
    "    # Separate corrected p-values for each dataframe\n",
    "    n_snps1 = len(df1.loc[0, \"p_values\"])\n",
    "    n_snps2 = len(df2.loc[0, \"p_values\"])\n",
    "    \n",
    "    corrected_p_values1 = corrected_p_values[:len(df1) * n_snps1]\n",
    "    corrected_p_values2 = corrected_p_values[len(df1) * n_snps1:]\n",
    "    \n",
    "    # Assign corrected p-values and significant SNPs\n",
    "    for i in range(len(df1)):\n",
    "        start = i * n_snps1\n",
    "        end = (i + 1) * n_snps1\n",
    "        df1.at[i, \"corrected_p_values\"] = corrected_p_values1[start:end]\n",
    "        df1.at[i, \"significant_snps\"] = {genotype_df.loc[snp_idx, 'Locus']: p \n",
    "                                         for snp_idx, p in enumerate(corrected_p_values1[start:end]) if p < 0.05}\n",
    "    \n",
    "    for i in range(len(df2)):\n",
    "        start = i * n_snps2\n",
    "        end = (i + 1) * n_snps2\n",
    "        df2.at[i, \"corrected_p_values\"] = corrected_p_values2[start:end]\n",
    "        df2.at[i, \"significant_snps\"] = {genotype_df.loc[snp_idx, 'Locus']: p \n",
    "                                         for snp_idx, p in enumerate(corrected_p_values2[start:end]) if p < 0.05}\n",
    "    \n",
    "    # Drop genes with no significant SNPs\n",
    "    eQTLs_df1 = df1[df1[\"significant_snps\"].map(lambda d: len(d)) > 0].reset_index(drop=True)\n",
    "    eQTLs_df2 = df2[df2[\"significant_snps\"].map(lambda d: len(d)) > 0].reset_index(drop=True)\n",
    "    \n",
    "    return eQTLs_df1, eQTLs_df2\n",
    "\n",
    "\n",
    "def eqtl_analysis(df1, df2, snps1, snps2):\n",
    "    eQTLs1 = association_test(df1, snps1)\n",
    "    eQTLs2 = association_test(df2, snps2)\n",
    "    eQTLs_corrected1, eQTLs_corrected2 = correct_results(eQTLs1, eQTLs2)\n",
    "    return eQTLs_corrected1, eQTLs_corrected2"
   ],
   "id": "26916a902260d30",
   "outputs": [],
   "execution_count": 16
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:44:09.081687Z",
     "start_time": "2024-09-20T14:42:01.869988Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Run the eQTL analysis\n",
    "eQTLs_myeloid, eQTLs_liver = eqtl_analysis(myeloid_df_filtered, liver_df_filtered, representative_snps_myeloid, representative_snps_liver)\n",
    "eQTLs_myeloid.head()"
   ],
   "id": "4b69f324f34ea5d5",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "            gene                                           p_values  \\\n",
       "0  0610009K11Rik  [0.7873410843132229, 0.8146248196203343, 0.903...   \n",
       "1  0710001D07Rik  [0.5113191855604085, 0.9841955580636499, 0.852...   \n",
       "2  1110001E17Rik  [0.42799892033109144, 0.698007729042269, 0.313...   \n",
       "3  1110003E01Rik  [0.9394831525677712, 0.9875195913130943, 0.585...   \n",
       "4  1110003F05Rik  [0.0062575741664354576, 0.02298229911840907, 0...   \n",
       "\n",
       "                                  corrected_p_values  \\\n",
       "0  [0.9936389945208776, 0.9945867590017438, 0.997...   \n",
       "1  [0.9816872876639764, 0.9995894401713452, 0.995...   \n",
       "2  [0.9772561651921646, 0.990235038035205, 0.9703...   \n",
       "3  [0.9984314843536239, 0.9996823519891541, 0.985...   \n",
       "4  [0.8568268500457448, 0.9154534596026077, 0.842...   \n",
       "\n",
       "                                    significant_snps  \n",
       "0  {'rs3667720': 0.03810062116258838, 'rs13476083...  \n",
       "1  {'rs3726974': 0.00017985910254448507, 'rs31438...  \n",
       "2               {'rs13476315': 0.026984285342278033}  \n",
       "3  {'rs13476118': 0.008429474007565099, 'rs134761...  \n",
       "4                   {'D1Mit294': 0.0487819157620498}  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_values</th>\n",
       "      <th>corrected_p_values</th>\n",
       "      <th>significant_snps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0610009K11Rik</td>\n",
       "      <td>[0.7873410843132229, 0.8146248196203343, 0.903...</td>\n",
       "      <td>[0.9936389945208776, 0.9945867590017438, 0.997...</td>\n",
       "      <td>{'rs3667720': 0.03810062116258838, 'rs13476083...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0710001D07Rik</td>\n",
       "      <td>[0.5113191855604085, 0.9841955580636499, 0.852...</td>\n",
       "      <td>[0.9816872876639764, 0.9995894401713452, 0.995...</td>\n",
       "      <td>{'rs3726974': 0.00017985910254448507, 'rs31438...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1110001E17Rik</td>\n",
       "      <td>[0.42799892033109144, 0.698007729042269, 0.313...</td>\n",
       "      <td>[0.9772561651921646, 0.990235038035205, 0.9703...</td>\n",
       "      <td>{'rs13476315': 0.026984285342278033}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1110003E01Rik</td>\n",
       "      <td>[0.9394831525677712, 0.9875195913130943, 0.585...</td>\n",
       "      <td>[0.9984314843536239, 0.9996823519891541, 0.985...</td>\n",
       "      <td>{'rs13476118': 0.008429474007565099, 'rs134761...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1110003F05Rik</td>\n",
       "      <td>[0.0062575741664354576, 0.02298229911840907, 0...</td>\n",
       "      <td>[0.8568268500457448, 0.9154534596026077, 0.842...</td>\n",
       "      <td>{'D1Mit294': 0.0487819157620498}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 18
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:44:13.587437Z",
     "start_time": "2024-09-20T14:44:13.569619Z"
    }
   },
   "cell_type": "code",
   "source": "eQTLs_liver.head()",
   "id": "b3299ff7e6c3c21d",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                                gene  \\\n",
       "0  1-acylglycerol-3-phosphate O-acyltransferase 5...   \n",
       "1     1-acylglycerol-3-phosphate O-acyltransferase 9   \n",
       "2      2',3'-cyclic nucleotide 3' phosphodiesterase\"   \n",
       "3                 2'-5' oligoadenylate synthetase 1C   \n",
       "4   3-hydroxy-3-methylglutaryl-Coenzyme A synthase 1   \n",
       "\n",
       "                                            p_values  \\\n",
       "0  [0.8231205823389012, 0.9384462933247342, 0.662...   \n",
       "1  [0.41408148322813854, 0.33823247465348305, 0.8...   \n",
       "2  [0.16864091640940057, 0.08769081945237145, 0.2...   \n",
       "3  [0.00010081372774950156, 0.000528171112483109,...   \n",
       "4  [0.513470486979324, 0.6258490256587621, 0.4778...   \n",
       "\n",
       "                                  corrected_p_values  \\\n",
       "0  [0.9948606669580963, 0.9984035097639842, 0.988...   \n",
       "1  [0.9764860823745213, 0.9717115103325055, 0.995...   \n",
       "2  [0.9585972268523969, 0.9465455399806717, 0.968...   \n",
       "3  [0.27395850104159125, 0.5582374436424771, 0.88...   \n",
       "4  [0.981810618738754, 0.9871111157664438, 0.9799...   \n",
       "\n",
       "                                    significant_snps  \n",
       "0  {'rs13476898': 0.019543743198308247, 'rs621910...  \n",
       "1  {'CEL-2_50605053': 0.0019472312833531023, 'rs1...  \n",
       "2           {'CEL-6_17260153': 0.030412039228340775}  \n",
       "3  {'rs6201380': 0.020337933483872327, 'gnf01.013...  \n",
       "4                {'rs3659850': 0.002031989164650012}  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_values</th>\n",
       "      <th>corrected_p_values</th>\n",
       "      <th>significant_snps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1-acylglycerol-3-phosphate O-acyltransferase 5...</td>\n",
       "      <td>[0.8231205823389012, 0.9384462933247342, 0.662...</td>\n",
       "      <td>[0.9948606669580963, 0.9984035097639842, 0.988...</td>\n",
       "      <td>{'rs13476898': 0.019543743198308247, 'rs621910...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1-acylglycerol-3-phosphate O-acyltransferase 9</td>\n",
       "      <td>[0.41408148322813854, 0.33823247465348305, 0.8...</td>\n",
       "      <td>[0.9764860823745213, 0.9717115103325055, 0.995...</td>\n",
       "      <td>{'CEL-2_50605053': 0.0019472312833531023, 'rs1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2',3'-cyclic nucleotide 3' phosphodiesterase\"</td>\n",
       "      <td>[0.16864091640940057, 0.08769081945237145, 0.2...</td>\n",
       "      <td>[0.9585972268523969, 0.9465455399806717, 0.968...</td>\n",
       "      <td>{'CEL-6_17260153': 0.030412039228340775}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2'-5' oligoadenylate synthetase 1C</td>\n",
       "      <td>[0.00010081372774950156, 0.000528171112483109,...</td>\n",
       "      <td>[0.27395850104159125, 0.5582374436424771, 0.88...</td>\n",
       "      <td>{'rs6201380': 0.020337933483872327, 'gnf01.013...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3-hydroxy-3-methylglutaryl-Coenzyme A synthase 1</td>\n",
       "      <td>[0.513470486979324, 0.6258490256587621, 0.4778...</td>\n",
       "      <td>[0.981810618738754, 0.9871111157664438, 0.9799...</td>\n",
       "      <td>{'rs3659850': 0.002031989164650012}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 19
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-20T14:44:15.436695Z",
     "start_time": "2024-09-20T14:44:15.429483Z"
    }
   },
   "cell_type": "code",
   "source": [
    "eQTLs_counter = Counter()\n",
    "for eqtl in pd.concat([eQTLs_myeloid['significant_snps'], eQTLs_liver['significant_snps']], axis=0):\n",
    "    eQTLs_counter.update(eqtl.keys())\n",
    "eqtls_number = len(eQTLs_counter)\n",
    "print(f\"There are {eqtls_number} eQTLs\")"
   ],
   "id": "d4b01dad3ee34a86",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1179 eQTLs\n"
     ]
    }
   ],
   "execution_count": 20
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "eQTLs_counter",
   "id": "77ec5b6e6d0b329f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "f0837350d074b51e",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
